### ISSUE TITLE
Windows: Fresh install of mlpack & simple project = MLPACK_ARMA_USE_OPENMP warning

### ISSUE THREAD

**Issue by DanL1234** [mlpack/mlpack#2971]
Not sure how to resolve this. 

I've gone through the download / installation procedure twice and get the same error for a fresh build. More precisely:

Windows 10 / vcpkg:

1. git clone https://github.com/microsoft/vcpkg (within c:\src\vcpkg\ folder via powershell)
2. navigate to c:\src\vcpkg\
3. .\bootstrap-vcpkg.bat
4. .\vcpkg install mlpack:x64-windows
5. .\vcpkg integrate install
6. open Visual Studio project and build all.

Visual Studio project is as simple as it can be:

#include <iostream>
#include "mlpack\core.hpp"
int main(int argc, char** argv)
{
    //some_timer starts ticking..
    mlpack::Timer::Start("some_timer");
    //some outputs.
    mlpack::Log::Debug << "Compiled with debugging symbols." << std::endl;
    mlpack::Log::Info << "Some test informational output." << std::endl;
    mlpack::Log::Warn << "A warning!" << std::endl;
    mlpack::Log::Warn << "Made it!" << std::endl;
    //some_timer stops..
    mlpack::Timer::Stop("some_timer");
    return 0;
}

Visual Studio build output:

Rebuild started...
1>------ Rebuild All started: Project: ConsoleApplication1, Configuration: Debug x64 ------
1>Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29914 for x64
1>Copyright (C) Microsoft Corporation.  All rights reserved.
1>cl /c /IC:\Projects\headers\ /I"C:\src\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\include" /ZI /JMC /W3 /WX- /diagnostics:column /sdl /Od /D _DEBUG /D _CONSOLE /D _UNICODE /D UNICODE /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /permissive- /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"x64\Debug\\" /Fd"x64\Debug\vc142.pdb" /Gd /TP /FC /errorReport:prompt ConsoleApplication1.cpp
1>ConsoleApplication1.cpp
1>C:\src\vcpkg\installed\x64-windows\include\mlpack\core\util\arma_config_check.hpp(54,1): warning C4081: expected '('; found 'string'
1>   Creating library C:\Users\tedil\source\repos\MLPack2_GitHubWikiDemo\ConsoleApplication1\x64\Debug\ConsoleApplication1.lib and object C:\Users\tedil\source\repos\MLPack2_GitHubWikiDemo\ConsoleApplication1\x64\Debug\ConsoleApplication1.exp
1>ConsoleApplication1.vcxproj -> C:\Users\tedil\source\repos\MLPack2_GitHubWikiDemo\ConsoleApplication1\x64\Debug\ConsoleApplication1.exe
1>Done building project "ConsoleApplication1.vcxproj".
========== Rebuild All: 1 succeeded, 0 failed, 0 skipped ==========

^ The line with the warning "arma_config_check.hpp" shows me the MLPACK_ARMA_USE_OPENMP message, which reads:

"mlpack was compiled with OpenMP support, but you are \
compiling without OpenMP support.  This will almost certainly cause \
irreparable disaster.  Either enable OpenMP support in your application (e.g., \
add -fopenmp to your compiler command line), or, recompile mlpack *without* \
OpenMP support."

Points to note:

1. If I add the brackets to the pragma message, the message scrolls out in the build results window. 
2. I said I did this twice, from scratch. In my first attempt at installing and building the above, I ended up passing (*) the /openmp- compile option for the vistual studio compiler so that openmp would not be built for. However, some mlpack dependencies look to use gcc instead of the visual studio standard compiler (cl.exe), which doesnt understand "/openmp-", so the build fails when it gets to those dependencies - and of course, I cannot then build my project on top of that.

(*) Nb. I did this within the vcpkg triplet x64-windows.cmake with,
"
set(VCPKG_CXX_FLAGS "/openmp- ${VCPKG_CXX_FLAGS}")
set(VCPKG_CXX_FLAGS_DEBUG "/openmp- ${VCPKG_CXX_FLAGS_DEBUG}")
set(VCPKG_CXX_FLAGS_RELEASE "/openmp- ${VCPKG_CXX_FLAGS_RELEASE}")
set(VCPKG_C_FLAGS "/openmp- ${VCPKG_C_FLAGS}")
set(VCPKG_C_FLAGS_DEBUG "/openmp- ${VCPKG_C_FLAGS_DEBUG}")
set(VCPKG_C_FLAGS_RELEASE "/openmp- ${VCPKG_C_FLAGS_RELEASE}")
"

3. Please note, although I have programmed C/C++ for years, I am really and truely a complete novice at cmake, vcpkg, package managers and the like. As such, I may be using not-quite correct terminology. Apologies in advance!!

4. I guess that if I can force vcpkg to always use cl.exe, I could build mlpack and all dependencies with or without /openmp. Unfortunately, after around 1 week of reading on the topic, I can find no way to do this. 

Besides the above number-4, I believe mplack and its dependencies should build without warning of disaster when freshly installed & compiled. 

Anywaysssss..... Please help? 

Thanks & kind regards,
Daniel

**Comment by shrit** [comment]
@DanL1234 Thanks for opening the issue with such details.
When I see the error message, I understand that you have already acquired mlpack binaries and installed them on your machine.

Are you sure that you have `openMP` installed at your machine?
If yes

Did you add `-fopenmp` to your compiler command line on your project?
It seems to me that you have tried to add the flags (I have never used Visual Studio, but I believe they are done correctly), 
and you mentioned that other mlpack dependencies are looking for gcc? Can you be more specific here?

I think you are referring to armadillo, and you are using the autodownloader that download all mlpack dependencies when you building mlpack from the scratch (correct me If I am wrong), I would suggest that you only download armadillo, and build it locally, then restart the process with mlpack.

Let me know if this works for you.

**Comment by DanL1234** [comment]
Hi Shrit, thanks for the response. I'm not able to look at this for a day or two but wanted to thank you and let you know I will look at your idea (building armadillo seperately sounds a great idea - just need to look into how to do that :-) ).

I'll come back to you shortly and let you know how I get along. Cheeeeers!!! :-D

**Comment by DanL1234** [comment]
Hi Omar, All,

Quote Omar, "Are you sure that you have openMP installed at your machine?  If yes Did you add -fopenmp to your compiler command line on your project?"

Honestly, I dont know how to "install" openMP - if I look at the openMP site, it's just a set of coding standards and interfaces. Visual Studio (uses cl.exe to compile/build) supports openmp via "\openmp" and "\openmp-" compiler options - and yes, I did ensure these were included (more details below)

So, an update: I have spent the last few weeks trying to dig into this issue, got sidetracked with a vcpkg issue where it appeared that two different compilers were being called (https://github.com/microsoft/vcpkg/issues/18423) 

In summary:

1. In the above issue, I had added C / CXX compile flags within the vcpkg x64-windows triplet before calling .\vcpkg install mlpack:x64-windows. This didnt work as although the flags were for cl.exe they ended up being passed to gcc when compiling lapack. Consequently, I had to try to figure out what was going on and how to work around it. (see linked bug above)
2. The workaround I have been trying is as I posted in the other issue linked above: create a fresh clone of vcpkg, bootstrap it, then install lapack (no compiler flags are set by me at this point - it's all a fresh clone). All goes smoothly for lapack. Next, set the C & CXX flags in vcpkg to either "/openmp-" or "/openmp", and call ".\vcpkg install mlpack:x64-windows". This again goes smoothly, this time with the C & CXX flags set to support openmp or not. 
3. Open t he visual studio project and set the openmp settings as per your choice of the C/CXX compile flags, and rebuild (ensuring the settings are in the right project setting: debug/release, and, x86/x64)

Nb. I repeated this same procedure from number-2 above on a fresh install with both "/openmp-" or "/openmp" - ie, trying both options.

In all cases I end up with an MLPACK_ARMA_ warning. I havent tried deleting the cached builds that vcpkg keeps, so not sure if this is fooling me. I would need to spend more time on it to figure it out.

Can anyone provide any more guidance? (Or thoughts/comments?)

Apologies / thanks in advance,
Dan

**Comment by JackBoosY** [comment]
The warning was written here:
https://github.com/mlpack/mlpack/blob/0d9de4b541de12413dcd77d184390a3a59b2cb59/src/mlpack/core/util/arma_config_check.hpp#L40-L56
So if `ARMA_USE_OPENMP` is not defined and `MLPACK_ARMA_USE_OPENMP` is defined, the compiler will report this error message.
And `ARMA_USE_OPENMP` is a macro of `armadillo`. The feature `openmp` is not enabled when armadillo is built by vcpkg, so when mlpack uses armadillo, this macro cannot be automatically passed to build mlpack.

At last, there is no option to enable openmp in the cmake build system of armadillo, so we currently cannot add openmp as a feature.

**Comment by rcurtin** [comment]
Hmm, I think that if vcpkg is not using OpenMP for its Armadillo build, then you could force mlpack to *not* compile with OpenMP support by specifying the CMake configuration flag `USE_OPENMP=OFF`.  Maybe I overlooked something that was written, but would that be a workable solution here to simply disable OpenMP support?

**Comment by JackBoosY** [comment]
> Hmm, I think that if vcpkg is not using OpenMP for its Armadillo build, then you could force mlpack to _not_ compile with OpenMP support by specifying the CMake configuration flag `USE_OPENMP=OFF`. Maybe I overlooked something that was written, but would that be a workable solution here to simply disable OpenMP support?

When I force to set `USE_OPENMP=ON`:
```
CMake Error in src/mlpack/CMakeLists.txt:
  Evaluation file to be written multiple times with different content.  This
  is generally caused by the content evaluating the configuration type,
  language, or location of object files:

   F:/vcpkg/buildtrees/mlpack/x64-windows-rel/src/mlpack/mlpack_CXX_Release_cotire.cmake


CMake Error in src/mlpack/CMakeLists.txt:
  Evaluation file to be written multiple times with different content.  This
  is generally caused by the content evaluating the configuration type,
  language, or location of object files:

   F:/vcpkg/buildtrees/mlpack/x64-windows-rel/src/mlpack/mlpack_CXX_Release_cotire.cmake
```

Any ideas?

**Comment by rcurtin** [comment]
I've never seen that issue before.  But my suggestion was to set `USE_OPENMP` to `OFF`, not `ON`, and see if that fixes the error.  Does that work to solve the problem?

**Comment by JackBoosY** [comment]
@rcurtin See my patch: 
[fix-build-with-openmp.zip](https://github.com/mlpack/mlpack/files/6819986/fix-build-with-openmp.zip)

I've successfully build mlpack with openmp without this changes.

**Comment by rcurtin** [comment]
Very nice!  Thank you for the patch!  A couple small questions before I open a PR with it though:

 - Are you sure it is a good idea to make finding OpenMP `REQUIRED`?  Personally I think that since `USE_OPENMP` defaults to `ON`, that we just choose to not use it if a user doesn't have it.  But maybe there is some reason I overlooked?

 - Are you sure it's necessary to remove the settings for `CMAKE_CXX_FLAGS` and `CMAKE_C_FLAGS`?  Or is that only a thing we need to avoid on Windows?

**Comment by JackBoosY** [comment]
> Are you sure it is a good idea to make finding OpenMP REQUIRED? Personally I think that since USE_OPENMP defaults to ON, that we just choose to not use it if a user doesn't have it. But maybe there is some reason I overlooked?

That's only suitable for vcpkg, please do not use it.

> Are you sure it's necessary to remove the settings for CMAKE_CXX_FLAGS and CMAKE_C_FLAGS? Or is that only a thing we need to avoid on Windows?

Yes, since openmp provided the target, we should use it instead of the c/cxx flags. `CMAKE_C_FLAGS` and `CMAKE_CXX_FLAGS` are only use for toolchain or buildsystem.

**Comment by DanL1234** [comment]
Hi All,

Thank you for attempting to help me with this issue. Unfortunately, the issue is not resolved from my side. 

Nb. In the following, I am not using any C or CXX flags.

### Num.1 - attempting to install mlpack via vcpkg.
. as above, calling ".\vcpkg install mlpack[openmp]:v64-windows" (see https://github.com/microsoft/vcpkg/pull/18942 ) - all dependencies downloaded afresh, built okay, no issues.
. as above, then calling ".\vcpkg integrate install" 
. then opening Visual Studio basic-a-project-that-can-possibly-be (!) and rebuild. 

> I still get the MLPACK_ARMA_USE_OPENMP warning within arma_config_check.hpp

### Num.2 - following the mlpack build/install instructions here: https://www.mlpack.org/doc/mlpack-3.4.2/doxygen/build_windows.html

**ie. the process being:**
. download mlpack (v3.4.2.), unzip to c:\mlpack\mlpack
. Visual Studio -> Project from existing code, located in c:\mlpack\mlpack
. Visual Studio -> ensure debug x64 is selected **AND** set the C/C++ language option for openmp (&) on all configurations.
. Visual Studio -> add OpenBlas via NuGet
. (I have boost binaries already, as directed - no issue here, just FYI)
. downloaded armadillo 10.6.1, unzip to c:\mlpack\armadillo, created .\build\ folder and ran cmake (details in link above)
. opened ".\build\armadillo.sln", set the C/C++ language option for openmp (&) on all configurations, then built the solution (no issues). Closed Visual Studio.
. Navigated to c:\mlpack\mlpack\build and ran cmake as described. (no issues)
. Opened the new mlpack.sln in the mlpack\build\ folder, set the C/C++ language option for openmp (&) on all configurations and build the solution. 

(&) - At these points, I set the C/C++ language option for openmp option within the Visual Studio solution. Initially I tried setting to not use openmp in all cases, then subsequently, retried and set to use openmp in all cases. Ie, ensuring all solutions are set the same way at each point in the process.

> I still get the MLPACK_ARMA_USE_OPENMP warning within arma_config_check.hpp


### Aside
It's also worth saying that there are many many issues with type defines:
"warning C4244: 'argument': conversion from 'const size_t' to 'const double', possible loss of data"
or with templates
"message : see reference to function template instantiation ... "
So much so that after a few thousand lines of text warning, I cease building.
(But this is another issue)


Can anyone replicate my issues?

**Comment by DanL1234** [comment]
@JackBoosY, you wrote:

> So if `ARMA_USE_OPENMP` is not defined and `MLPACK_ARMA_USE_OPENMP` is defined, the compiler will report this error message.

okay...

> And `ARMA_USE_OPENMP` is a macro of `armadillo`. The feature `openmp` is not enabled when armadillo is built by vcpkg, so when mlpack uses armadillo, this macro cannot be automatically passed to build mlpack.

yep, so now I do not use vcpkg to build armadillo.

> `MLPACK_ARMA_USE_OPENMP` is automatic set to `ON` when openmp was found.
> So this issue happened.

Okay, so it seems from text in the cmake files that this is "detected" from the compiler capability, not from a projects setting, as per my (&) references in the posts above.

> At last, there is no option to enable openmp in the cmake build system of armadillo, so we currently cannot add openmp as a feature.

As I understand the situation, the cmake build system creates the Visual Studio solution - which I then open and ensure it uses the C/C++ language setting for openmp - set to on or off, depending on which attempt we talk about - so, I think I get-around the cmake issue (?). 

I guess, in other-words, what I'm trying to say is that you tell me that cmake build system for armadillo cannot offer openmp settings (currenlty), but, the output of the cmake system is the armadillo.sln file, which I can manually set the openmp setting as I wish. My thoughts being, I can set armadillo and mlpack solutions the same way. Yet, still I end up with arma_config.hpp defining the MLPACK_ARMA_USE_OPENMP warning.

**Comment by JackBoosY** [comment]
Was this issue fixed?

**Comment by conradsnicta** [comment]
TLDR: use of OpenMP under MSVC is not advised.

Longer explanation: Armadillo requires a compiler with support for the OpenMP 3.1 standard.  Last time I checked, MSVC only supports the comically ancient OpenMP 2.0 standard, which lacks certain critical features.

See the following for more details:  
https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.6.x/include/armadillo_bits/compiler_setup.hpp#L372-409

**Comment by DanL1234** [comment]
Thanks Jack & Conrad.

Was it fixed? Not to my knowledge or understanding. Whatever path I chose, whatever compiler settings I used, I always ended up with the same MLPACK_ARMA_USE_OPENMP warning when building mlpack. 

My fix? Okay, it's a workaround, but I pulled all armadillo files (wrappers & hpps) into the same project as the mlpack source and compiled them together (to create one executable, rather than separate libs). Doing this, I still got the warning in the arma_config / arma_config_check files, but this is a legacy of the cmake system creating those files & warnings, so I felt safe in ignoring / deleting the warnings. 

After all, one compilation of all files cant result in compilation differences and similar issues, right!!

It seems to work as I built my first CNN on MNIST dataset the other night. 97% accuracy!! (Happy with that)

Thanks for following up.

**Comment by DanL1234** [comment]
Ps. I had to move on and find an alternative solution (for me) as I couldnt spend any more time on it. That's why I havent commented here until now. Sorry about that; would've loved to have been able to help.

**Comment by JackBoosY** [comment]
This issue is in the build system, not the code.

**Comment by rcurtin** [comment]
As a follow-up, we are in the process of transitioning mlpack to be a header-only library. When that happens, we can always use whatever Armadillo's detected settings for OpenMP (and other options) are. This makes packaging a lot easier and I think it should mean issues like this one will no longer be possible. :+1:
